bson/src/main/org/bson/BsonArray.java

```diff
@@ -208,4 +206,23 @@ public String toString() {
                + "values=" + values
                + '}';
     }
+
+    @Override
+    public BsonArray clone() {
+        BsonArray to = new BsonArray();
+        for (BsonValue cur : this) {
+            if (cur.isDocument()) {
```

Comment:
Not sure how important this is, but doesn't BsonValue have an enum property that indicates what type it is? And if so, wouldn't a switch statement be much faster? There isn't any way to optimize this loop in the jvm, so using a jump table seems like a good idea on a critical path.